docs/proposals/let-function.rst

```diff
@@ -0,0 +1,218 @@
+lexical scope.  This will require some changes to the lookup semantics in
+JMESPath to introduce scoping, but provides useful functionality such as being
+able to refer to elements defined outside of the current scope used to evaluate
+an expression.
+
+
+Motivation
+==========
+
+As a JMESPath expression is being evaluated, the current element, which can be
+explicitly referred to via the ``@`` token, changes as expressions are
+evaluated.  Given a simple sub expression such as ``foo.bar``, first the
+``foo`` expression is evaluted with the starting input JSON document, and the
+result of that expression is then used as the current element when the ``bar``
+element is evaluted.  Conceptually we're taking some object, and narrowing down
+its current element as the expression is evaluted.
+
+Once we've drilled down to a specific current element, there is no way, in the
+context of the currently evaluated expression, to refer to any elements outside
+of that element.  One scenario where this is problematic is being able to refer
+to a parent element.
+
+For example, suppose we had this data::
+
+  {"first_choice": "WA",
+   "states": [
+     {"name": "WA", "cities": ["Seattle", "Bellevue", "Olympia"]},
+     {"name": "CA", "cities": ["Los Angeles", "San Francisco"]},
+     {"name": "NY", "cities": ["New York City", "Albany"]},
+   ]
+  }
+
+Let's say we wanted to get the list of cities of the state corresponding to our
+``first_choice`` key.  We'll make the assumption that the state names are
+unique in the ``states`` list.  This is currently not possible with JMESPath.
+In this example we can hard code the state ``WA``::
+
+  states[?name==`WA`].cities
+
+but it is not possible to base this on a value of ``first_choice``, which
+comes from the parent element.  This JEP proposes a solution that makes
+this possible in JMESPath.
+
+
+Specification
+=============
+
+There are two components to this JEP, a new function, ``let()``, and a change
+to the way that identifiers are resolved.
+
+
+The let() Function
+------------------
+
+The ``let()`` function is heavily inspired from the ``let`` function commonly
+seen in the Lisp family of languages:
+
+* https://clojuredocs.org/clojure.core/let
+* http://docs.racket-lang.org/guide/let.html
+
+The let function is defined as follows::
+
+  any let(object scope, expression->any expr)
+
+``let`` is a function that takes two arguments.  The first argument is a JSON
+object.  This hash defines the names and their corresponding values that will
+be accessible to the expression specified in the second argument.  The second
+argument is an expression reference that will be evaluated.
+
+Resolving Identifiers
+---------------------
+
+Prior to this JEP, identifiers are resolved by consulting the current context
+in which the expression is evaluted.  For example, using the same
+``search`` function as defined in the JMESPath specification, the
```

Comment:
This part threw me off, some a minor point of feedback might be to link to what you're talking about. I thought `search` was a JMESPath function.